glean-core/src/upload/mod.rs

```diff
@@ -30,49 +30,82 @@ use util::process_pings_dir;
 mod request;
 mod util;
 
+/// The state of the upload manager queue at the time of asking.
+#[derive(PartialEq, Debug)]
+pub enum PingsQueueStatus {
```

Comment:
Not very sure about naming choices here, regarding the enum and its options. Would appreciate some better suggestions if anyone has them :)

Reply:
What about describing what the receiver needs to do:
It's an `UploadJob` / `UploadTask`:

* `Upload(PingRequest)` or `Send` - as your `Next`
* `Wait(seconds)` -  Wait the amount of seconds before asking again (the receiver doesn't care why it needs to wait. This will then also apply once we add throttling.
* `None`/`Nothing`/`Done` - No more job/task, end your work.

What do you think?

Reply:
Paging in @Dexterp37 for some bikeshedding.

Reply:
I'll wait for @Dexterp37 's opinions before settling this, but I will comment on the `Wait(seconds)`. I do agree that will make sense for throttling, but I didn't add it here this time because I didn't find it necessary yet.

Do you think I should add it now? And if so, how many seconds should I default to?

Reply:
> 
> 
> What about describing what the receiver needs to do:
> It's an `UploadJob` / `UploadTask`:
> 
>     * `Upload(PingRequest)` or `Send` - as your `Next`

+1 on `Upload`
 
>     * `Wait(seconds)` -  Wait the amount of seconds before asking again (the receiver doesn't care why it needs to wait. This will then also apply once we add throttling.

+1 on this!

>     * `None`/`Nothing`/`Done` - No more job/task, end your work.
> 
> 
> What do you think?

I wonder if we even need this third one. @brizental when will the product use the APIs? Do you expect the APIs to be called right after startup (only once, then governed by the `Wait(seconds)` for the rest of the time)? Or do you expect the product to poll at specified times?

Reply:
We do need a third one. The idea was for the upload job on the platform side to ask for jobs and upload them and only stop working when instructed.

Consider what happens to pending pings at startup, say 3:
* On Glean.init we start an upload worker, that one asks the API for something to upload
* It gets 3 jobs, one at a time, then finishes

Reply:
> I wonder if we even need this third one. @brizental when will the product use the APIs? Do you expect the APIs to be called right after startup (only once, then governed by the Wait(seconds) for the rest of the time)? Or do you expect the product to poll at specified times?

This is an interesting question. I imagined the upload worker would be called at exaclty the same times as it is called right now (on init, whenever pings are submitted). I guess we do have the ability to change it up here and just give the platform a timeframe as to "when it should come back", so that the UploadWorker would be started on init and just told to come back every given interval.

This raises questions that I don't know how to answer such as: when **should** it come back?

Now, taking a step back, if we stick to what I initially was certain of: that the upload worker will ask for "next pings" on init and whenever it submits new pings we do need the `None` option. Because the platform needs to know when to stop asking for pings.

If we change it up to some other flow, it seems reasonable to me that we still keep the `None` option, for the same reason. But I am not so sure about it.

Reply:
I don't think we need to change where the upload worker is started.
Just right now we don't explicitly start one to look for pending pings from previous runs. This just happens as a side effect of other pings (e.g. startup metrics pings or similar).

I'm fine with leaving out `Wait` for now until we actually implement backoff.

Reply:
> Now, taking a step back, if we stick to what I initially was certain of: that the upload worker will ask for "next pings" on init and whenever it submits new pings, we do need the `None` option. Because the platform needs to know when to stop asking for pings.

Perfect, the above makes sense to me :-) Let's retain the `None` option and leave `Wait` for when the backoff is implemented.